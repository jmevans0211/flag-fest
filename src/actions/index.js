export const saveCountries = countries => ({
  type: 'SAVE_COUNTRIES',
  countries
});