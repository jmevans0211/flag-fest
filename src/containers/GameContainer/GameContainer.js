import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { saveCountries, removeCountryGuessed } from './../../actions';
import FlagCard from '../FlagCard/FlagCard';
import ResultsCard from './../../components/ResultsCard/ResultsCard';

export class GameContainer extends Component {
  constructor() {
    super();
    this.state = {
      points: 0,
      flagsGuessed: [],
    }
  }


  handleGuess = (answer) => {
    const { countries, removeCountryGuessed } = this.props
    if(answer === 'correct') {
      this.addPoints()
      this.setState({ flagsGuessed: [...this.state.flagsGuessed, countries[0]]})
      removeCountryGuessed();
    } else if (answer === 'incorrect') {
      this.setState({ flagsGuessed: [...this.state.flagsGuessed, countries[0]]})
      removeCountryGuessed();
    }
  }

  addPoints = () => {
    this.setState({ points: this.state.points += 1})
  }
  
  render() {
   const { countries } = this.props

    return (
      <main>
        {countries.length !== 0 && <FlagCard handleGuess={this.handleGuess} flagsGuessed={this.state.flagsGuessed} />}
        {countries.length === 0 && <ResultsCard points={this.state.points} flagsGuessed={this.state.flagsGuessed} />}
        {countries.length !== 0 && 
          <Link to="/">
            <p>Start Over</p>
          </Link>
        }
      </main>
    )
  }
}

export const mapStateToProps = state => ({
  countries: state.countries,
});

export const mapDispatchToProps = dispatch => ({
  saveCountries: countries => dispatch(saveCountries(countries)),
  removeCountryGuessed: () => dispatch(removeCountryGuessed())
});

export default connect(mapStateToProps, mapDispatchToProps)(GameContainer);


