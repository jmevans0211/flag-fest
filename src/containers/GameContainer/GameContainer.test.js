import React from 'react';
import { shallow } from 'enzyme';
import { GameContainer, mapStateToProps, mapDispatchToProps } from './GameContainer';
import { saveCountries, removeCountryGuessed } from '../../actions';

describe ('GameContainer', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<GameContainer countries={
      [
        {
          flag: 'https://restcountries.eu/data/srb.svg',
          name: 'Serbia',
          numericCode: 688,
          region: 'Europe'
        },
        {
          flag: 'https://restcountries.eu/data/lux.svg',
          name: 'Luxembourg',
          numericCode: 442,
          region: 'Europe'
        }
      ]
    }
    correctClass = 'positionA'
    wrongClass = 'positionB'
    />)
  });

  it.skip('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it.skip('should match snapshot when class names change for answer buttons', () => {
    const wrapper = shallow(
      <GameContainer countries={
        [
          {
            flag: 'https://restcountries.eu/data/srb.svg',
            name: 'Serbia',
            numericCode: 688,
            region: 'Europe'
          },
          {
            flag: 'https://restcountries.eu/data/lux.svg',
            name: 'Luxembourg',
            numericCode: 442,
            region: 'Europe'
          }
        ]
      }
      correctClass = 'positionB'
      wrongClass = 'positionA'
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should update state of points when addPoints is called', () => {
    const expected = 1;

    expect(wrapper.state('points')).toEqual(0);

    wrapper.instance().addPoints();

    expect(wrapper.state('points')).toEqual(expected);
  });

});

describe('mapStateToProps', () => {
  it('should return an object with a countries array', () => {
    const mockState = {
      countries: [
        {
          flag: 'https://restcountries.eu/data/srb.svg',
          name: 'Serbia',
          numericCode: 688,
          region: 'Europe'
        },
        {
          flag: 'https://restcountries.eu/data/lux.svg',
          name: 'Luxembourg',
          numericCode: 442,
          region: 'Europe'
        }
      ],
      filter: 'SAVE_COUNTRIES'
    };
    const expected = {
      countries: mockState.countries
    }

      const mappedProps = mapStateToProps(mockState);

      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    let mockCountries;
    let mockDispatch;
    beforeEach(() => {
      mockCountries = [
        {
          flag: 'https://restcountries.eu/data/srb.svg',
          name: 'Serbia',
          numericCode: 688,
          region: 'Europe'
        },
        {
          flag: 'https://restcountries.eu/data/lux.svg',
          name: 'Luxembourg',
          numericCode: 442,
          region: 'Europe'
        }
      ]
      mockDispatch = jest.fn();
    })
    it('should get countries based on region', () => {
      const actionToDispatch = saveCountries('SAVE_COUNTRIES', mockCountries);
      const mappedProps = mapDispatchToProps(mockDispatch);
  
      mappedProps.saveCountries('SAVE_COUNTRIES', mockCountries)
  
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should delete country from store', () => {
      const actionToDispatch = removeCountryGuessed('REMOVE_COUNTRY_GUESSED');
      const mappedProps = mapDispatchToProps(mockDispatch)
      
      mappedProps.removeCountryGuessed('REMOVE_COUNTRY_GUESSED');

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });