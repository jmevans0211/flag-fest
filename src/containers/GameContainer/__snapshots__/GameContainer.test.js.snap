// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameContainer should match snapshot 1`] = `
<main>
  <Connect(FlagCard)
    handleGuess={[Function]}
  />
  <Link
    to="/"
  >
    <p>
      Start Over
    </p>
  </Link>
</main>
`;
